### PR TITLE
Colorado Low income child care expense credit

### LINKED ISSUE #2724: Colorado Low income child care expense credit
https://tax.colorado.gov/sites/tax/files/documents/DR_0104_Book_2022.pdf 
Page 48

https://casetext.com/statute/colorado-revised-statutes/title-39-taxation/specific-taxes/income-tax/article-22-income-tax/part-1-general/section-39-22-1195-child-care-expenses-tax-credit-legislative-declaration-definitions (3)

### PR DESCRIPTION
**Pull request by Jingj-WANG**
Fixes #2724

Co-authored-by: Pavel Makarchuk <PavelMakarchuk@users.noreply.github.com>

### Summary
📄🧒🌄

This pull request adds two YAML files that define the parameters of the low income child care expense credit in Colorado, a state-level tax benefit for working families. The files are `policyengine_us/tests/policy/baseline/gov/states/co/tax/income/credits/cdcc/low_income/max_amount.yaml` and `policyengine_us/tests/policy/baseline/gov/states/co/tax/income/credits/cdcc/low_income/rate.yaml`.

> _Adding `CO_LICCEC`_
> _A new credit for child care_
> _Winter brings relief_

### Walkthrough
*  Add the low income child care expense credit in Colorado ([link](https://github.com/PolicyEngine/policyengine-us/pull/2765/files?diff=unified&w=0#diff-fdc4fed27027b75247b5482d4bd0cecd128904db90696bdcb0ebfc8e0f10702aR1-R20), [link](https://github.com/PolicyEngine/policyengine-us/pull/2765/files?diff=unified&w=0#diff-e715c4cfcf411ba88e755da97eec1726be2330c5ca96ed54433915c6e767e351L1-R21))
  - Define the maximum amount of the credit based on the number of dependents in `max_amount.yaml` ([link](https://github.com/PolicyEngine/policyengine-us/pull/2765/files?diff=unified&w=0#diff-fdc4fed27027b75247b5482d4bd0cecd128904db90696bdcb0ebfc8e0f10702aR1-R20))
  - Define the rate of the credit based on the federal adjusted gross income in `rate.yaml` ([link](https://github.com/PolicyEngine/policyengine-us/pull/2765/files?diff=unified&w=0#diff-e715c4cfcf411ba88e755da97eec1726be2330c5ca96ed54433915c6e767e351L1-R21))

### CODE REVIEW

File: policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/child_age_limit
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/child_age_limit:1]: need .yaml behind every file

File: policyengine_us/tests/policy/baseline/gov/states/co/tax/income/credits/cdcc/low_income/low_income_cdcc.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/co/tax/income/credits/cdcc/low_income/low_income_cdcc.yaml:1]: need more tests for edge cases

Comment (PavelMakarchuk) [reply policyengine_us/tests/policy/baseline/gov/states/co/tax/income/credits/cdcc/low_income/low_income_cdcc.yaml:1]: test fails

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    def formula(tax_unit, spm_unit, period, parameters):
+        p = parameters(period).gov.states.co.tax.income.credits.cdcc.low_income
+        # Filer is eligible if AGI is below $25,000
+        agi = tax_unit("adjusted_gross_income", period)
+        # The childcare expenses can not exceed the earned income of the filer
+        # if filing jointly - the lesser of either individual's earned income
+        earned_income = tax_unit("min_head_spouse_earned", period)
+        # Get sum of all qualified child care expenses
+        expenses = spm_unit("childcare_expenses", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:21]:
```suggestion
        expenses = tax_unit.spm_unit("childcare_expenses", period)
```

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        person = spm_unit.members
+        age = person("age", period)
+        age_eligible = age < p.child_age_limit
+        eligible_expenses = age_eligible * expenses
+        total_expenses = spm_unit.sum(eligible_expenses)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:26]:
```suggestion
        total_expenses = tax_unit.sum(eligible_expenses)
```

File: policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/child_age_limit.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+description: Colorado provides low-income family child care expenses credit if child age are below this limit
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/child_age_limit.yaml:1]:
```suggestion
description: Colorado qualifies children for the low-income family child care expenses credit below this age threshold.
```

File: policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/max_amount.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+description: Colorado provides the following low income child care expense credit amount, based on the number of dependents.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/max_amount.yaml:1]:
```suggestion
description: Colorado allows fro the following low income child care expense credit maximum amount, based on the number of dependents.
```

File: policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/child_age_limit.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+metadata:
+  reference:
+    - title: 39-22-119.5. Child care expenses tax credit - legislative declaration - definitions. (3)(a)(III)
+      href: https://advance.lexis.com/documentpage/?pdmfid=1000516&crid=19fe6818-36b1-4893-af3b-666480efcb35&config=014FJAAyNGJkY2Y4Zi1mNjgyLTRkN2YtYmE4OS03NTYzNzYzOTg0OGEKAFBvZENhdGFsb2d592qv2Kywlf8caKqYROP5&pddocfullpath=%2Fshared%2Fdocument%2Fstatutes-legislation%2Furn%3AcontentItem%3A61P5-WYK1-DYDC-J4SD-00008-00&pdcontentcomponentid=234176&pdteaserkey=sr1&pditab=allpods&ecomp=7s65kkk&earg=sr1&prid=cafebc4f-f2fc-4bd5-85bb-13a2b27d5e40
+    - title: 2022 Colorado Individual Income Tax Filing Guide, page 46, Eligible, 4, b
+      href: https://tax.colorado.gov/sites/tax/files/documents/DR_0104_Book_2022.pdf#page=46 
+    - title: 2021 Colorado Individual Income Tax Filing Guide, page 44, Eligible, 4, b
+      href: https://tax.colorado.gov/sites/tax/files/documents/DR_0104_Book_2022.pdf#page=44  
+
+  unit: years
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/child_age_limit.yaml:13]:
```suggestion
  unit: year
```

File: policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/rate.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+metadata:
+  type: single_amount
+  threshold_unit: currency-USD
+  amount_unit: /1
+  label: Colorado low income child care expense credit rate
+  reference:
+    - title: Colorado revised statutes, Section 39-22-119.5 - Child care expenses tax credit (3), (a), (I)
+      href: https://casetext.com/statute/colorado-revised-statutes/title-39-taxation/specific-taxes/income-tax/article-22-income-tax/part-1-general/section-39-22-1195-child-care-expenses-tax-credit-legislative-declaration-definitions
+    - title: 2022 Colorado Individual Income Tax Filing Guide, page 48
+      href: https://tax.colorado.gov/sites/tax/files/documents/DR_0104_Book_2022.pdf
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/rate.yaml:11]: need pages in these links as well

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/rate.yaml:11]: in all yaml files

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        person = tax_unit.members
+        age = person("age", period)
+        age_eligible = age < p.child_age_limit
+        eligible_expenses = age_eligible * expenses
+        total_expenses = tax_unit.sum(eligible_expenses)
+        capped_expenses = min_(earned_income, total_expenses)
+        co_cdcc = capped_expenses * p.rate.calc(agi)
+        # Credit is capped at $500 or $1,000 for 1 or over 2 dependents respectfully
+        # dependents = tax_unit("tax_unit_dependents", period)
+        spm_unit_count_children = add(tax_unit, period, ["is_child"])
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:31]:
```suggestion
        count_children = add(tax_unit, period, ["is_child"])
```

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        p = parameters(period).gov.states.co.tax.income.credits.cdcc.low_income
+        # Filer is eligible if AGI is below $25,000
+        agi = tax_unit("adjusted_gross_income", period)
+        # The childcare expenses can not exceed the earned income of the filer
+        # if filing jointly - the lesser of either individual's earned income
+        earned_income = tax_unit("min_head_spouse_earned", period)
+        # Get sum of all qualified child care expenses
+        expenses = tax_unit.spm_unit("childcare_expenses", period)
+        person = tax_unit.members
+        age = person("age", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:23]: check for "is_child" here and make 
child = person("is_child", period)
age_eligible = (age < p.child_age_limit) & is_child

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py (hunk 7 above)
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:31]:
```suggestion
```

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        age_eligible = age < p.child_age_limit
+        eligible_expenses = age_eligible * expenses
+        total_expenses = tax_unit.sum(eligible_expenses)
+        capped_expenses = min_(earned_income, total_expenses)
+        co_cdcc = capped_expenses * p.rate.calc(agi)
+        # Credit is capped at $500 or $1,000 for 1 or over 2 dependents respectfully
+        # dependents = tax_unit("tax_unit_dependents", period)
+        spm_unit_count_children = add(tax_unit, period, ["is_child"])
+        max_amount = p.max_amount.calc(spm_unit_count_children)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:32]:
```suggestion
        max_amount = p.max_amount.calc(tax_unit.sum(eligible_child))
```

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py (hunk 7 above)
Comment (Jingj-WANG) [reply policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:31]: don't understand what is comment means

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py (hunk 9 above)
Comment (Jingj-WANG) [reply policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:32]: the test file cannot pass if using "tax_unit.sum(eligible_child)", but can pass using "spm_unit_count_children" in calc(). 
image

In "spm_unit_count_children = add(tax_unit, period, ["is_child"])", add doesn't mean sum????

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:32]: we want to count only the eligible children (under 13) for the max amount, not all children, that's why the adjustment

Comment (Jingj-WANG) [reply policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:33]: got it. but it is strange that using "max_amount = p.max_amount.calc(spm_unit_count_children)" can pass the test. let me check the test file also.

Comment (Jingj-WANG) [reply policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:33]: oh, I understand why. Will add another example in the test file

**Comment by Jingj-WANG** [comment]
Two Questions: 

1. we are checking the children's age, and whether we need to change "person = tax_unit.members" to "person = tax_unit.spm_unit.members"
2. In the threshold file, it only contains the lower limit, not the upper one, so when there's only one child who meets the requirements, it should be 500 instead of 1000, I changed the upper limit to 1.1. and test passed. Not sure whether it's correct.

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:1]: We have to include a condition that hey cant get the normal cdcc
image

File: policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/max_amount.yaml [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+      href: https://tax.colorado.gov/sites/tax/files/documents/DR_0104_Book_2022.pdf#page=48 
+    - title: 2021 Colorado Individual Income Tax Filing Guide, page 46
+      href: https://tax.colorado.gov/sites/tax/files/documents/DR0104Book_2021_V3.pdf#page=46
+brackets:
+  - threshold:
+      2018-01-01: 0
+    amount:
+      2018-01-01: 500
+  - threshold:
+      2018-01-01: 1.1
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/max_amount.yaml:20]:
```suggestion
      2018-01-01: 1
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/max_amount.yaml:20]: lower bracket is exclusive

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/rate.yaml
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/rate.yaml:1]: In general we favor this compactness, but because the law defines the AGI limit in such a separate part from the match, I'd suggest breaking this into two parameters. Then make a `co_low_income_cdcc_eligible` variable which is based on the four provisions of 3(a) in the law.

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        child = person("is_child", period)
+        eligible_child = (age < p.child_age_threshold) & child
+        eligible_expenses = eligible_child * expenses
+        total_expenses = tax_unit.sum(eligible_expenses)
+        capped_expenses = min_(earned_income, total_expenses)
+        co_low_income_cdcc = capped_expenses * p.rate.calc(agi)
+        # Credit is capped at $500 or $1,000 for 1 or over 2 dependents respectfully
+        eligible_children = tax_unit.sum(eligible_child)
+        max_amount = p.max_amount.calc(eligible_children)
+        # Filer is not eligible for the low income cdcc if he received the CO cdcc
```
[the remark this answers is not in the document]
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:34]: I read it as specifying that the credit under [39-22-119](https://law.justia.com/codes/colorado/2016/title-39/specific-taxes/article-22/part-1/section-39-22-119) - which is the co cdcc - I might be wrong, then we can use the tax before non-refundable (tax form says tax from federal income tax return can not greater than 0)

image

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc_eligible.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc_eligible.py:1]: need a unit test for this variable

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    defined_for = "co_low_income_cdcc_eligible"
+
+    def formula(tax_unit, period, parameters):
+        p = parameters(period).gov.states.co.tax.income.credits.cdcc.low_income
+        # The childcare expenses can not exceed the earned income of the filer
+        # if filing jointly - the lesser of either individual's earned income
+        earned_income = tax_unit("min_head_spouse_earned", period)
+        # Get sum of all qualified child care expenses
+        spm_unit = tax_unit.spm_unit
+        spm_unit_cdcc_expense = spm_unit("childcare_expenses", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py:20]:
```suggestion
        childcare_expense = spm_unit("childcare_expenses", period)
```

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc.py]: I agree, let's change it to `co_tax_before_non_refundable_credits > 0`

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc_eligible.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    definition_period = YEAR
+    defined_for = StateCode.CO
+
+    def formula(tax_unit, period, parameters):
+        p = parameters(period).gov.states.co.tax.income.credits.cdcc.low_income
+        # Filer is eligible if AGI is below $25,000
+        agi = tax_unit("adjusted_gross_income", period)
+        agi_eligible = agi <= p.income_threshold
+        # Filers cannot claim both the low income CDCC and the CO CDCC
+        co_cdcc = tax_unit("co_cdcc", period) > 0
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc_eligible.py:18]:
```suggestion
        co_cdcc = tax_unit("co_tax_before_non_refundable_credits", period) > 0
```

File: policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/rate.yaml [hunk 14]
```diff
@@ -0,0 +1,14 @@
+description: Colorado provides the following low income child care expense credit rate.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/co/tax/income/credits/cdcc/low_income/rate.yaml:1]:
```suggestion
description: Colorado matches up to this fraction of qualifying child care expenses in its low income child care expenses credit.
```

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc_eligible.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+    documentation = "https://casetext.com/statute/colorado-revised-statutes/title-39-taxation/specific-taxes/income-tax/article-22-income-tax/part-1-general/section-39-22-1195-child-care-expenses-tax-credit-legislative-declaration-definitions"
+    definition_period = YEAR
+    defined_for = StateCode.CO
+
+    def formula(tax_unit, period, parameters):
+        p = parameters(period).gov.states.co.tax.income.credits.cdcc.low_income
+        # Filer is eligible if AGI is below $25,000
+        agi = tax_unit("adjusted_gross_income", period)
+        agi_eligible = agi <= p.income_threshold
+        # Filers cannot claim both the low income CDCC and the CO CDCC
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc_eligible.py:17]:
```suggestion
        # Filers cannot claim the low income CDCC if they don't have tax liability to
        # claim the non-refundable Colorado CDCC.
```

File: policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc_eligible.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+        tax_before_non_refundable_credits = (
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/co/tax/income/credits/cdcc/low_income/co_low_income_cdcc_eligible.py:18]:
```suggestion
        owes_tax_before_non_refundable_credits = (
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

**Review by MaxGhenis — APPROVED** [review]
